azuki-core/src/main/kotlin/com/anaplan/engineering/azuki/core/scenario/VerifiableScenarioCursor.kt

```diff
@@ -0,0 +1,105 @@
+package com.anaplan.engineering.azuki.core.scenario
+
+import com.anaplan.engineering.azuki.core.system.*
+import org.slf4j.LoggerFactory
+
+class VerifiableScenarioCursor<AF : ActionFactory, CF : CheckFactory>(
+    private val systemFactory: SystemFactory<AF, CF, *, *, *>,
+    private val scenario: VerifiableScenario<AF, CF>,
+) {
+    private val systemCursor by lazy {
+        val declarations = scenario.declarations(systemFactory.actionFactory)
+        if (declarations.filterIsInstance<UnsupportedAction>().isNotEmpty()) {
+            Log.warn("Unsupported declaration found! Declarations: ${declarations.joinToString("\n") { "\t - $it" }}")
+            state = State.UnsupportedDeclaration
+            null
+        } else {
+            val systemDefinition = SystemDefinition(declarations)
+            when (systemFactory) {
+                is MutableSystemFactory -> MutableSystemCursor(systemFactory.create(systemDefinition))
+                else -> ImmutableSystemCursor(systemFactory, systemDefinition)
+            }
+        }
+    }
+
+    var state = State.Ok
+        private set
+
+    enum class State {
+        Ok,
+        UnsupportedDeclaration,
+        UnsupportedCommand,
+        NoChecks,
+    }
+
+    private val scenarioIterations = scenario.iterations().toMutableList()
+
+    val isEmpty get() = scenarioIterations.isEmpty()
+
+    fun hasNext(): Boolean {
+        if (state != State.Ok || isEmpty) {
+            return false
+        }
+        val lookAhead = scenarioIterations.first()
+        if (UnsupportedAction in lookAhead.commands(systemFactory.actionFactory)) {
+            state = State.UnsupportedCommand
+            return false
+        }
+        if (lookAhead.checks(systemFactory.checkFactory).all { it == UnsupportedCheck }) {
+            state = State.NoChecks
+            return false
+        }
+        return true
+    }
+
+    fun next(): System<AF, CF> {
+        if (!hasNext()) {
+            throw IllegalStateException("Cursor cannot be moved")
+        }
+        val scenarioIteration = scenarioIterations.removeFirst()
+        val systemIteration = SystemIteration(
+            commands = scenarioIteration.commands(systemFactory.actionFactory),
+            checks = scenarioIteration.checks(systemFactory.checkFactory),
+            regardlessOfActions = scenarioIteration.regardlessOfActions(systemFactory.actionFactory)
+        )
+        return systemCursor!!.apply(systemIteration)
+    }
+
+    fun destroy() = systemCursor?.destroy()
+
+    companion object {
+        private val Log = LoggerFactory.getLogger(VerifiableScenarioCursor::class.java)
+    }
+}
+
+sealed interface SystemCursor<AF : ActionFactory, CF : CheckFactory> {
+    fun apply(iteration: SystemIteration): System<AF, CF>
+    fun destroy() {}
+}
+
+class MutableSystemCursor<AF : ActionFactory, CF : CheckFactory>(val system: MutableSystem<AF, CF>) :
```

Comment:
We should almost certainly apply the same cursor/mutable system pattern for action generators